clients/rospy/src/rospy/impl/registration.py

```diff
@@ -306,6 +306,8 @@ def run(self):
                 if cond is not None:
                     cond.release()
 
+            get_topic_manager().check_all()
```

Comment:
Now that I rewrote the check function I'd say it is not expensive. The lock is now only acquired when one or more connections need to be removed. The only overhead that we add now is the check for connections that need to be closed which is O(1) with epoll/kqueue (with poll it would be O(n)!).